sockeye/inference.py

```diff
@@ -880,25 +980,33 @@ def _beam_search(self,
 
         return sequences, attentions, scores_accumulated, lengths
 
-    @staticmethod
-    def _get_best_from_beam(sequences: mx.nd.NDArray,
+    def _get_best_from_beam(self,
+                            sequences: mx.nd.NDArray,
                             attention_lists: mx.nd.NDArray,
                             accumulated_scores: mx.nd.NDArray,
-                            lengths: mx.nd.NDArray) -> Translation:
+                            lengths: mx.nd.NDArray) -> List[Translation]:
         """
         Return the best (aka top) entry from the n-best list.
 
-        :param sequences: Array of word ids. Shape: (beam_size, bucket_key).
+        :param sequences: Array of word ids. Shape: (num_sent * beam_size, bucket_key).
         :param attention_lists: Array of attentions over source words. Shape: (length, bucket_key).
         :param accumulated_scores: Array of length-normalized negative log-probs.
-        :return: Top sequence, top attention matrix, top accumulated score (length-normalized negative log-probs)
-                 and length.
+        :return: Top sequence, top attention matrix, top accumulated score (length-normalized
+                 negative log-probs) and length.
         """
+        #utils.check_condition(sequences.shape[0] == attention_lists.shape[0] \
+        #                      == accumulated_scores.shape[0] == lengths.shape[0], "Shape mismatch")
+
         # sequences & accumulated scores are in latest 'k-best order', thus 0th element is best
         best = 0
-        length = int(lengths[best].asscalar())
-        sequence = sequences[best][:length].asnumpy().tolist()
-        # attention_matrix: (target_seq_len, source_seq_len)
-        attention_matrix = np.stack(attention_lists[best].asnumpy()[:length, :], axis=0)
-        score = accumulated_scores[best].asscalar()
-        return Translation(sequence, attention_matrix, score)
+        num_sent = sequences.shape[0] // self.beam_size
+        result = []
+        for sent in range(num_sent):
+            idx = sent * self.beam_size + best
+            length = int(lengths[idx].asscalar())
+            sequence = sequences[idx].asnumpy().tolist()
```

Comment:
why do you not cut to length here anymore?

Reply:
Good catch! Was lost in refactoring.